apps/survey/contracts/Survey.sol

```diff
@@ -0,0 +1,250 @@
+            // index 0 in options array is reserved for NO_VOTE, so here we are comparing
+            // the option just added (_options[j] = options[j+1]) with the previous one (options[j])
+            require(survey.voters[msg.sender].castedVotes[j].optionId < survey.voters[msg.sender].castedVotes[j + 1].optionId);
+            totalVoted = totalVoted.add(_stakes[j]);
+
+            CastVote(_surveyId, msg.sender, _options[j], _stakes[j]);
+        }
+
+        // compute and register non used tokens
+        // implictly we are doing require(totalVoted <= voterStake) too
+        // (as stated before, index 0 is for NO_VOTE option)
+        survey.voters[msg.sender].castedVotes[0].stake = voterStake.sub(totalVoted);
+
+        // register number of options voted
+        survey.voters[msg.sender].optionsCastedLength = _options.length;
+
+        // add voter tokens to participation
+        survey.participation = survey.participation.add(totalVoted);
+    }
+
+    /**
+    * @notice Vote option #`_option` in survey #`_surveyId`.
+    * @dev It will use the whole balance. If voting option is 0, that's like cancelling the previously casted vote
+    * @param _surveyId Id for survey
+    * @param _option Index of supported option
+    */
+    function voteOption(uint256 _surveyId, uint256 _option) public {
+        SurveyStruct storage survey = surveys[_surveyId];
+        // this could re-enter, though we can asume the governance token is not maliciuous
+        uint256 voterStake = token.balanceOfAt(msg.sender, survey.snapshotBlock);
+        uint256[] memory options = new uint256[](1);
+        uint256[] memory stakes = new uint256[](1);
+        options[0] = _option;
+        stakes[0] = voterStake;
+
+        voteOptions(_surveyId, options, stakes);
+    }
+
+    function canVote(uint256 _surveyId, address _voter) public view returns (bool) {
+        SurveyStruct storage survey = surveys[_surveyId];
+
+        return _isSurveyOpen(survey) && token.balanceOfAt(_voter, survey.snapshotBlock) > 0;
+    }
+
+    function getSurvey(uint256 _surveyId) public view returns (bool open, address creator, uint64 startDate, uint256 snapshotBlock, uint256 minParticipationPct, uint256 votingPower, uint256 participation, uint256 options) {
+        SurveyStruct storage survey = surveys[_surveyId];
+
+        open = _isSurveyOpen(survey);
+        creator = survey.creator;
+        startDate = survey.startDate;
+        snapshotBlock = survey.snapshotBlock;
+        minParticipationPct = survey.minAcceptParticipationPct;
+        votingPower = survey.votingPower;
+        participation = survey.participation;
+        options = survey.options;
+    }
+
+    function getSurveyMetadata(uint256 _surveyId) public view returns (string) {
+        return surveys[_surveyId].metadata;
+    }
+
+    function getVoterState(uint256 _surveyId, address _voter) public view returns (uint256[] options, uint256[] stakes) {
```

Comment:
Wondering if this should be make `external` instead.

I think we keep most of the getters public so that inherited contracts can use them too (usually as a convenience), but I'm worried about this one because it has a potentially unbounded for-loop.

An inherited contract has access to read this data using `surveys[_surveyId].voters[_voter].castedVotes` (and they probably don't need the entire array duplicated for them when they do want to read some part of it).

To me, this is something that just shouldn't be used internally.